### PR TITLE
Added proxy user + pw detection

### PR DESCRIPTION
**Pull request by patrickskowronekdkfz**
Currently the config allows to use the user and password of a proxy as var or encoded in the url. The added function provides a possibility to cover both use cases

### CODE REVIEW

File: lib/update-bridgehead.sh [hunk 1]
```diff
@@ -55,7 +55,7 @@ for DIR in /etc/bridgehead $(pwd); do
     OUT=$(retry 5 git -C $DIR fetch 2>&1 && retry 5 git -C $DIR pull 2>&1)
   else
     log "INFO" "Git is using proxy ${HTTP_PROXY_URL} from ${CONFFILE}"
-    OUT=$(retry 5 git -c http.proxy=$HTTP_PROXY_URL -c https.proxy=$HTTPS_PROXY_URL -C $DIR fetch 2>&1 && retry 5 git -c http.proxy=$HTTP_PROXY_URL -c https.proxy=$HTTPS_PROXY_URL -C $DIR pull 2>&1)
+    OUT=$(retry 5 git -c http.proxy=$PROXY -c https.proxy=$PROXY -C $DIR fetch 2>&1 && retry 5 git -c http.proxy=$PROXY -c https.proxy=$PROXY -C $DIR pull 2>&1)
```
Comment (torbrenner) [inline lib/update-bridgehead.sh:58]: You are mixing http and https proxy. This will lead to https calls beeing send to the http proxy, which could matter if the sites have two different proxies for that. I think we will need to have the code above create the PROXY variable for http and a SECURE_PROXY variable for https

**Review by torbrenner — CHANGES_REQUESTED** [review]

File: lib/functions.sh [hunk 2]
```diff
@@ -3,12 +3,28 @@
 detectCompose() {
 	if [[ "$(docker compose version 2>/dev/null)" == *"Docker Compose version"* ]]; then
 		COMPOSE="docker compose"
-	else
+	e
```
Comment (lablans) [inline lib/functions.sh:6]: Even if this works, please don't do that

File: lib/functions.sh [hunk 3]
```diff
@@ -3,12 +3,28 @@
...
+		PROXY="$(echo $PROTO$HTTP_PROXY_USERNAME:$HTTP_PROXY_PASSWORD@$URL)"
+
+		SECURE_PROTO="$(echo $HTTPS_PROXY_URL | grep :// | sed -e's,^\(.*://\).*,\1,g')"
+		SECURE_URL="$(echo ${HTTPS_PROXY_URL/$SECURE_PROTO/})"
+		SECURE_PROXY="$(echo $SECURE_PROTO$HTTP_PROXY_USERNAME:$HTTP_PROXY_PASSWORD@$SECURE_URL)"
+	else
+		PROXY=$HTTP_PROXY_URL
+		SECURE_PROXY=$HTTPS_PROXY_URL
+	fi
+}
```
Comment (lablans) [inline lib/functions.sh:26]: Please consider other variable names, these are quite ambiguous.

**Review by lablans — CHANGES_REQUESTED** [review]

File: lib/functions.sh (hunk 2 above)
Comment (patrickskowronekdkfz) [reply lib/functions.sh]: Fixed. I don't recall ever changing this var.

File: lib/prerequisites.sh [hunk 4]
```diff
@@ -68,7 +68,7 @@ source /etc/bridgehead/${PROJECT}.conf
 source ${PROJECT}/vars
 
 set +e
-SERVERTIME="$(https_proxy=$HTTPS_PROXY_URL curl -m 5 -s -I $BROKER_URL_FOR_PREREQ 2>&1 | grep -i -e '^Date: ' | sed -e 's/^Date: //i')"
+SERVERTIME="$(https_proxy=$HTTPS_PROXY_FULL_URL curl -m 5 -s -I $BROKER_URL_FOR_PREREQ 2>&1 | grep -i -e '^Date: ' | sed -e 's/^Date: //i')"
```
Comment (TKussel) [inline lib/prerequisites.sh:71]:
```suggestion
SERVERTIME="$(http_proxy=$HTTP_PROXY_FULL_URL https_proxy=$HTTPS_PROXY_FULL_URL curl -m 5 -s -I $BROKER_URL_FOR_PREREQ 2>&1 | grep -i -e '^Date: ' | sed -e 's/^Date: //i')"
```
The broker is not required to run under HTTPS

File: lib/functions.sh
Comment (TKussel) [inline lib/functions.sh:1]: This change does not pick up the case of an `HTTP[S]?_PROXY_USERNAME` without password. I don't know if that's legal in HTTP, though.

File: lib/monitoring.sh [hunk 5]
```diff
@@ -47,8 +47,8 @@ function hc_send(){
 
     if [ -n "$2" ]; then
         MSG="$2\n\nDocker stats:\n$UPTIME"
-        echo -e "$MSG" | https_proxy=$HTTPS_PROXY_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null -X POST --data-binary @- "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
+        echo -e "$MSG" | https_proxy=$HTTPS_PROXY_FULL_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null -X POST --data-binary @- "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
```
Comment (TKussel) [inline lib/monitoring.sh:50]:
```suggestion
        echo -e "$MSG" | http_proxy=$HTTP_PROXY_FULL_URL https_proxy=$HTTPS_PROXY_FULL_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null -X POST --data-binary @- "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
```
While our healthcheck is served via HTTPS, this is not required.

File: lib/monitoring.sh [hunk 6]
```diff
@@ -47,8 +47,8 @@ function hc_send(){
...
     else
-        https_proxy=$HTTPS_PROXY_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
+        https_proxy=$HTTPS_PROXY_FULL_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
```
Comment (TKussel) [inline lib/monitoring.sh:52]:
```suggestion
       http_proxy=$HTTP_PROXY_FULL_URL  https_proxy=$HTTPS_PROXY_FULL_URL curl --max-time 5 -A "$USER_AGENT" -s -o /dev/null "$HCURL"/"$1" || log WARN "Monitoring failed: Unable to send data to $HCURL/$1"
```

File: lib/update-bridgehead.sh [hunk 7]
```diff
@@ -50,12 +50,11 @@ for DIR in /etc/bridgehead $(pwd); do
     git -C $DIR config credential.helper "$CREDHELPER"
   fi
   old_git_hash="$(git -C $DIR rev-parse --verify HEAD)"
-  if [ -z "$HTTP_PROXY_URL" ]; then
-    log "INFO" "Git is using no proxy!"
+  if [ -z "$HTTP_PROXY_FULL_URL" ]; then
```
Comment (TKussel) [inline lib/update-bridgehead.sh:53]: This is wrong, if `HTTPS_PROXY_FULL_URL` is set, but `HTTP_PROXY_FULL_URL` is not!

File: lib/update-bridgehead.sh [hunk 8]
```diff
@@ -50,12 +50,11 @@ for DIR in /etc/bridgehead $(pwd); do
     git -C $DIR config credential.helper "$CREDHELPER"
   fi
   old_git_hash="$(git -C $DIR rev-parse --verify HEAD)"
-  if [ -z "$HTTP_PROXY_URL" ]; then
-    log "INFO" "Git is using no proxy!"
```
Comment (TKussel) [inline lib/update-bridgehead.sh:54]: Why did you remove that?

**Review by TKussel — CHANGES_REQUESTED** [review]

File: lib/functions.sh
Comment (lablans) [reply lib/functions.sh:1]: agreed -- it would be technically legal but so uncommon that it doesn't matter

File: lib/monitoring.sh (hunk 5 above)
Comment (lablans) [reply lib/monitoring.sh:50]: Service is in fact required to run under https, cf. data protection concept(s)

File: lib/prerequisites.sh (hunk 4 above)
Comment (lablans) [reply lib/prerequisites.sh:71]: Service is in fact required to run under https, cf. data protection concept(s)

File: lib/monitoring.sh (hunk 6 above)
Comment (lablans) [reply lib/monitoring.sh:52]: Service is in fact required to run under https, cf. data protection concept(s)

File: lib/update-bridgehead.sh (hunk 8 above)
Comment (lablans) [reply lib/update-bridgehead.sh:54]: Thanks, fixed in 9fc8564e4ef280403deb472a40eda255e02d740e

File: lib/update-bridgehead.sh (hunk 7 above)
Comment (lablans) [reply lib/update-bridgehead.sh]: thanks, fixed in 9fc8564e4ef280403deb472a40eda255e02d740e

**Comment by TKussel** [comment]
I commented out the usage of an HTTP_PROXY as it is forbidden by our data protection concepts

**Review by TKussel — APPROVED** [review]

**Review by lablans — APPROVED** [review]
